### PR TITLE
Show stdout and stderr to course staff when evaluation doesn't pass

### LINKED ISSUE #1629: Show stdout and stderr of programming evaluation to course staff
Especially when the evaluation fails and there is no test report.

This will help staff to debug their student's code.

### PR DESCRIPTION
**Pull request by fonglh**
Fixes #1629.

screen shot 2016-11-02 at 5 14 17 pm

### CODE REVIEW

File: app/views/course/assessment/answer/programming/_programming.html.slim [hunk 1]
```diff
@@ -10,3 +10,5 @@ div.files
 - answer = last_attempt ? last_attempt.actable: programming
 = render partial: 'course/assessment/answer/programming/test_cases',
          locals: { question: question, answer: answer }
+= render partial: 'course/assessment/answer/programming/output_streams',
```
Comment (weiqingtoh) [inline app/views/course/assessment/answer/programming/_programming.html.slim]: Shall we move the checks from the partial into here instead? So we don't render the partial unless required.

```ruby
- auto_grading = answer.auto_grading ? answer.auto_grading.actable : nil
- is_grader = can?(:read_tests, @submission)
```

File: app/services/course/assessment/answer/programming_auto_grading_service.rb [hunk 2]
```diff
@@ -125,6 +126,20 @@ def build_failed_test_case_records(question, auto_grading)
...
 
+  # Sets results which belong to the auto grading rather than an individual test case.
+  #
+  # @param [Course::Assessment::Answer::ProgrammingAutoGrading] auto_grading The programming auto
+  #   grading result to store the test results in.
+  # @param [Course::Assessment::ProgrammingEvaluationService::Result] evaluation_result The
+  #   result of evaluating the package.
+  # @return [Course::Assessment::Answer::ProgrammingAutoGrading]
+  def set_auto_grading_results(auto_grading, evaluation_result)
+    auto_grading.stdout = evaluation_result.stdout
```
Comment (weiqingtoh) [inline app/services/course/assessment/answer/programming_auto_grading_service.rb]: Not a must, but can use `.tap` here.

**Comment by weiqingtoh** [comment]
Other than the check on the partial, LGTM!

**Comment by weiqingtoh** [comment]
Actually the [panel that you added in](https://github.com/Coursemology/coursemology2/pull/1630/files#diff-ebda536df353d12cc3128a5b3fb08b27R6) could be a common component that is usable across other parts of Coursemology. 

I was half thinking whether to abstract it out, to prevent others from having to repeat the same set of HTML code, but I'm not sure if we can do it as a react component instead. 

Thoughts? @allenwq @fonglh @kxmbrian

File: app/services/course/assessment/answer/programming_auto_grading_service.rb (hunk 2 above)
Comment (fonglh) [inline app/services/course/assessment/answer/programming_auto_grading_service.rb]: Like

```auto_grading.tap do |ag|
  ag.stdout = evaluation_result.stdout
  ag.stderr = evaluation_result.stderr
  ag.exit_code = evaluation_result.exit_code
end
```

Comment (weiqingtoh) [inline app/services/course/assessment/answer/programming_auto_grading_service.rb]: yup 👍

**Comment by allenwq** [comment]
@weiqingtoh React Component is a frontend thing, but here the template is rendered by backend, so not very suitable in this case. 

I think we can extract out when other components actually need this, that time we will also have a better sense what interface to provide.

File: app/services/course/assessment/answer/programming_auto_grading_service.rb (hunk 2 above)
Comment (fonglh) [inline app/services/course/assessment/answer/programming_auto_grading_service.rb]: Fixed

File: app/assets/stylesheets/course/assessment/submission/answer/answer.scss [hunk 3]
```diff
@@ -10,4 +10,17 @@
       margin-top: -15px;
     }
   }
+
+  .panel-heading .collapse-toggle::after {
+    // symbol for "opening" panels
+    color: $grey;
+    content: '\e114';
+    float: right;
+    font-family: 'Glyphicons Halflings';  // essential for enabling glyphicon
```
Comment (fonglh) [inline app/assets/stylesheets/course/assessment/submission/answer/answer.scss]: Switched to fontawesome chevrons.

**Pull request merged**